### ISSUE TITLE
Render dies out without error during custom BSDF evaluation

### ISSUE THREAD

**Issue by osylum** [mitsuba-renderer/mitsuba3#546]
## Summary

Render dies out without error during custom BSDF evaluation

## System configuration

System information:

OS: Windows-10
CPU: Intel64 Family 6 Model 165 Stepping 5, GenuineIntel
GPU: NVIDIA RTX A4000
Python: 3.9.7 (tags/v3.9.7:1016ef3, Aug 30 2021, 20:19:38) [MSC v.1929 64 bit (AMD64)]
NVidia driver: 517.40
CUDA: 10.0.130
LLVM: -1.-1.-1

Dr.Jit: 0.4.0
Mitsuba: 3.2.0
Is custom build? False
Compiled with: MSVC 19.34.31937.0
Variants:
scalar_rgb
scalar_spectral
cuda_ad_rgb
llvm_ad_rgb

## Description

I have a custom BSDF. Below is a minimal dummy implementation with some dummy variables just to reproduce the problem. I am using cuda_ad_rgb variant. When calling the eval() function it works. But when I try to render, the renders just dies out without error. It works by changing the line "theta_h = dr.acos(hz)" in the eval() function by -dr.pi. It will also die out when adding print for a variable (eg. print('h',h)). I would like to know what am doing wrong here, how to get get more info on the error, and how to print out information on the variables. Thank you

## Steps to reproduce

Below is only the BSDF dummy class. You can render it with any basic scene code. Let me know if you need a complete .py example?

<pre><code id="python_code">
class BSDF_Tabulated_Test(mi.BSDF):
    """
    tabulated brdf data
    """
    
    
    def __init__(self, props):
        mi.BSDF.__init__(self, props)
                
        brdf_table_raw = np.random.rand(90,90,180,3) # dummy table
        self.brdf_table = mi.TensorXf(brdf_table_raw.ravel().astype('float32'), shape=(90,90,180,3))
                                

        reflection_flags   = mi.BSDFFlags.SpatiallyVarying|mi.BSDFFlags.DiffuseReflection|mi.BSDFFlags.FrontSide | mi.BSDFFlags.BackSide
        self.m_components  = [reflection_flags]
        self.m_flags = reflection_flags

    def sample(self, ctx, si, sample1, sample2, active):
        
        #print('sample si: ', si)
        cos_theta_i = mi.Frame3f.cos_theta(si.wi)

        active &= cos_theta_i > 0

        bs = mi.BSDFSample3f()
        bs.wo  = mi.warp.square_to_cosine_hemisphere(sample2)
        bs.pdf = mi.warp.square_to_cosine_hemisphere_pdf(bs.wo)
        bs.eta = 1.0
        bs.sampled_type = +mi.BSDFFlags.DiffuseReflection
        bs.sampled_component = 0

        reflectance = mi.Vector3f(0.5)
        #value = self.m_reflectance.eval(si, active)
        value = reflectance

        #return ( bs, ek.select(active & (bs.pdf > 0.0), value, Vector3f(0)) )
        return ( bs, dr.select(active & (bs.pdf > 0.0), value, mi.Vector3f(0)) )
    
    
    def eval(self, ctx, si, wo, active):
        
        print('eval tabulated brdf')
        
        ind_min = 0
        ind_max = 89 - 1

        h = dr.normalize(si.wi + wo)
        hz = h.z
        hz = dr.clip(hz, 0., 1.) # should not be needed, but just in case
        if False: # turn off for success
            print('h',h)
        if True: # turn off for success
            theta_h = dr.acos(hz)
        else:
            theta_h = -dr.pi
        theta_h = dr.select(theta_h < 0., 0., theta_h)

        # mapping
        ind = ind_max * dr.sqrt(theta_h / (dr.pi * 0.5))
        ind = mi.Int(dr.floor(ind))
        # clamping
        ind = dr.select(ind < ind_min, ind_min, ind)
        ind = dr.select(ind > ind_max, ind_max , ind)
        
        print('evaluating table')
        val = self.brdf_table[ind, 0, 0, 0].array
        print('done evaluating table')
        return mi.Point3f(val)
    

    def pdf(self, ctx, si, wo, active):
        pdf = dr.clamp(mi.Frame3f.cos_theta(wo),1e-5,1)*1/dr.pi
        return pdf

    def eval_pdf(self, ctx, si, wo, active):
        f = self.eval(ctx,si,wo,active)
        pdf = self.pdf(ctx,si,wo,active)
        return f,pdf

    def parameters_changed(self, keys):
        pass

    def to_string(self):
        return ('Tabulated BRDF')  
</code></pre>

**Comment by osylum** [comment]
I don't see any error message, it seems to exit before print('done evaluating table') and does not render anything.

**Comment by njroussel** [comment]
I've just ran you script. Something must be wrong with your environment, I indeed get the following error message:
```
RuntimeError: jit_var_gather(): cannot gather from a placeholder variable!
```
The problem is the `val = self.brdf_table[ind, 0, 0, 0].array` line. The indexing creates a temporary/placeholder variable which you cannot read from anymore. Instead, you  should always `gather/scatter` from/to tensors directly. 
In your case, something like this:
```
lookup_idx = ind * dr.prod(self.brdf_table.shape[1:])
val = dr.gather(mi.Point3f, self.brdf_table.array, lookup_idx)
```
(You might want to double check the indexing, I didn't pay too much attention to what you were trying to access exactly)
